azurecompute/src/main/java/org/jclouds/azurecompute/compute/extensions/AzureComputeSecurityGroupExtension.java

```diff
@@ -0,0 +1,460 @@
+                  // TODO
+               }
+            }
+         }
+
+      }
+      return transformNetworkSecurityGroupToSecurityGroup(id);
+   }
+
+   @Override
+   public SecurityGroup addIpPermission(IpProtocol protocol, int startPort, int endPort,
+                                        Multimap<String, String> tenantIdGroupNamePairs,
+                                        Iterable<String> ipRanges,
+                                        Iterable<String> groupIds, SecurityGroup group) {
+      IpPermission.Builder permBuilder = IpPermission.builder();
+      permBuilder.ipProtocol(protocol);
+      permBuilder.fromPort(startPort);
+      permBuilder.toPort(endPort);
+      permBuilder.tenantIdGroupNamePairs(tenantIdGroupNamePairs);
+      permBuilder.cidrBlocks(ipRanges);
+      permBuilder.groupIds(groupIds);
+
+      return addIpPermission(permBuilder.build(), group);
+   }
+
+   @Override
+   public SecurityGroup removeIpPermission(IpPermission ipPermission, SecurityGroup group) {
+      checkNotNull(group, "group");
+      checkNotNull(ipPermission, "ipPermission");
+      String id = checkNotNull(group.getId(), "group.getId()");
+
+      String ruleName = NetworkSecurityGroups.createRuleName(ipPermission.getFromPort(), ipPermission.getToPort());
+      // remove rule to NSG
+      removeRuleFromNetworkSecurityGroup(id, ruleName);
+
+      // TODO remove endpoint from VM
+      Set<Deployment> deployments = FluentIterable.from(api.getCloudServiceApi().list())
+              .transform(new Function<CloudService, Deployment>() {
+                 @Override
+                 public Deployment apply(CloudService cloudService) {
+                    return api.getDeploymentApiForService(cloudService.name()).get(cloudService.name());
+                 }
+              })
+              .filter(Predicates.notNull())
+              .toSet();
+      // TODO filter deployments
+      for (Deployment deployment : deployments) {
+         Deployment.VirtualIP virtualIP = Iterables.tryFind(deployment.virtualIPs(), Predicates.notNull()).orNull();
+         if (virtualIP == null) throw new IllegalStateException("");
+
+         for (Role role : deployment.roles()) {
+            for (Role.ConfigurationSet configurationSet : role.configurationSets()) {
+               for (int i = ipPermission.getFromPort(); i <= ipPermission.getToPort(); i++) {
+                  String name = String.format(AzureComputeConstants.TCP_FORMAT, i, i);
+                  configurationSet.inputEndpoints().remove(createInputEndpoint(name, ipPermission.getIpProtocol()
+                          .name().toLowerCase(), virtualIP.address(), i));
+               }
+               String updateRoleRequestId = api.getVirtualMachineApiForDeploymentInService(deployment.name(), deployment.name()).updateRole(role.roleName(), role);
+               if (!operationSucceeded.apply(updateRoleRequestId)) {
+                  // TODO
```

Comment:
Throw an exception here to fail? Also, implement the other similar TODOs in this class, to have all execution paths covered.